### PR TITLE
Use IndexOfAnyExcept in BigInteger.IsPowerOfTwo

### PR DESCRIPTION
**Pull request by Rob-Hague**
This path is taken once we know the most significant uint is itself a power of two, and we just need to check the rest are zero. The benchmarks are an attempt to show the worst-case:

|             Method |        Job |      Mean |     Error |    StdDev | Ratio | RatioSD |
|------------------- |----------- |----------:|----------:|----------:|------:|--------:|
|    SmallPowerOfTwo | main |  1.436 ns | 0.0010 ns | 0.0009 ns |  1.00 |    0.00 |
|    SmallPowerOfTwo | PR |  2.503 ns | 0.0469 ns | 0.0392 ns |  1.74 |    0.03 |
|                    |            |           |           |           |       |         |
| SmallNotPowerOfTwo | main |  1.117 ns | 0.0012 ns | 0.0011 ns |  1.00 |    0.00 |
| SmallNotPowerOfTwo | PR |  2.464 ns | 0.0063 ns | 0.0059 ns |  2.21 |    0.01 |
|                    |            |           |           |           |       |         |
|    LargePowerOfTwo | main | 31.979 ns | 0.0081 ns | 0.0072 ns |  1.00 |    0.00 |
|    LargePowerOfTwo | PR |  5.711 ns | 0.0028 ns | 0.0025 ns |  0.18 |    0.00 |
|                    |            |           |           |           |       |         |
| LargeNotPowerOfTwo | main |  1.119 ns | 0.0011 ns | 0.0009 ns |  1.00 |    0.00 |
| LargeNotPowerOfTwo | PR |  2.919 ns | 0.0011 ns | 0.0009 ns |  2.61 |    0.00 |

```c#
public class Benchmarks
{
    private BigInteger _smallPowerOfTwo;
    private BigInteger _smallNotPowerOfTwo;
    private BigInteger _largePowerOfTwo;
    private BigInteger _largeNotPowerOfTwo;

    [GlobalSetup]
    public void GlobalSetup()
    {
        _smallPowerOfTwo    = BigInteger.Parse("0000000000000001000000000000000000000000000000000000000000000000", NumberStyles.BinaryNumber);
        _smallNotPowerOfTwo = BigInteger.Parse("0000000000000001000000000000000001110011011011000010100011101111", NumberStyles.BinaryNumber);

        Random rnd = new Random(123456);
        
        var bytes = new byte[1 + 256];
        bytes[0] = 1;
        
        _largePowerOfTwo = new BigInteger(bytes, isBigEndian: true);
        
        rnd.NextBytes(bytes.AsSpan(1));

        _largeNotPowerOfTwo = new BigInteger(bytes, isBigEndian: true);
    }

    [Benchmark]
    public bool SmallPowerOfTwo() => _smallPowerOfTwo.IsPowerOfTwo;
    
    [Benchmark]
    public bool SmallNotPowerOfTwo() => _smallNotPowerOfTwo.IsPowerOfTwo;
    
    [Benchmark]
    public bool LargePowerOfTwo() => _largePowerOfTwo.IsPowerOfTwo;
    
    [Benchmark]
    public bool LargeNotPowerOfTwo() => _largeNotPowerOfTwo.IsPowerOfTwo;
}
```

### CODE REVIEW

**Comment by danmoseley** [comment]
Thanks. Are the regressions worth it?

File: src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs [hunk 1]
```diff
@@ -637,15 +637,10 @@ public bool IsPowerOfTwo
...
-                if (!BitOperations.IsPow2(_bits[iu]))
-                    return false;
-                while (--iu >= 0)
-                {
-                    if (_bits[iu] != 0)
-                        return false;
-                }
-                return true;
+
+                return BitOperations.IsPow2(_bits[iu]) && _bits.AsSpan().IndexOfAnyExcept(0u) == iu;
```
Comment (huoyaoyuan) [inline src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs:643]:
```suggestion
                return BitOperations.IsPow2(_bits[iu]) && !_bits.AsSpan(0, iu).ContainsOfAnyExcept(0u);
```

Can it be better with `ContainsOfAnyExcept`?

Comment (Rob-Hague) [reply src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs]: `ContainsAnyExcept` currently calls to `IndexOfAnyExcept`, but if anything it is marginally better (no worse), and probably shows clearer intent. Thanks

**Comment by Rob-Hague** [comment]
> Thanks. Are the regressions worth it?

IndexOfAnyExcept (now ContainsAnyExcept) is "obviously" better in lots of cases, so I focused the benchmarks on where it isn't. I don't feel strongly about the change, but my thinking is:

1. If you have some code which is checking for powers of two, then there is *a priori* some likelihood that the input domain contains powers of two, in which case this is likely to be an improvement.
2. If the upper 32 bits of a value is already a power of two (popcount == 1), as we have determined before this path, then the likelihood is further increased for this particular value.
3. If someone cares about the performance in the small (~64 bit) case, then they should rather be using e.g. ulong or UInt128.

**Review by stephentoub — APPROVED** [review]

**Pull request merged**